### PR TITLE
Conditional application of InjectableArguments

### PR DESCRIPTION
**Pull request by kevin-montrose**
This is a proposal (and thus a draft PR) to add a `Condition` to behaviors that allows control of whether a behavior's `InjectableArguments` apply, akin to the Condition on attributes from #127 .

Currently, at least as I understand, it is possible to conditionally control what taint is applied to a returned value (either absolutely, or via arguments).  `InjectableArguments` are unconditionally considered.

The motivating use here are RedirectAction returning methods in the Stack Overflow codebase which default to disallowing external domains, but have a parameter that when set allows external targets.  Calls are therefore vulnerable to an open redirect injection only when the parameter is set.  A minimal example of this is in the [included test](https://github.com/kevin-montrose/security-code-scan/blob/1fb2b0d21e9c5f627fdbc32e570fc05266219e77/SecurityCodeScan.Test/Tests/Taint/OpenRedirectAnalyzerTest.cs#L333).

This also pulls in the commits from #134 , since that was needed to test successfully against the full Stack Overflow code base.

### CODE REVIEW

**Comment by kevin-montrose** [comment]
@JarLob I'll move this to be against `master` too, though probably not until Monday (need to test it against actual Stack Overflow, and that's tricky on the weekend).  If you have any comments before then, I can address them as part of moving these commits over as well.

File: SecurityCodeScan/Analyzers/Taint/CodeEvaluation/CSharpCodeEvaluation.cs [hunk 1]
```diff
@@ -778,6 +783,70 @@ private bool CheckPrecondition(IReadOnlyDictionary<int, object> condition, bool
...
+            if (condition == null || condition.Count == 0)
+                return true;
+
+            if (methodSymbol == null)
+                return false;
+
+            if (args == null)
+                return false;
+
+            var ps = methodSymbol.GetParameters();
```
Comment (JarLob) [inline SecurityCodeScan/Analyzers/Taint/CodeEvaluation/CSharpCodeEvaluation.cs]: This is not so commonly used extension wrapper. In this case it is safe to always assume `IMethodSymbol`. Just use `.Parameters`.

File: SecurityCodeScan/Analyzers/Taint/CodeEvaluation/CSharpCodeEvaluation.cs [hunk 2]
```diff
@@ -778,6 +783,70 @@ private bool CheckPrecondition(IReadOnlyDictionary<int, object> condition, bool
...
+
+            var vals = new string[ps.Length];
```
Comment (JarLob) [inline SecurityCodeScan/Analyzers/Taint/CodeEvaluation/CSharpCodeEvaluation.cs]: The function will be called often, during analysis of every configured function call. Ideally it should be allocation free.

File: SecurityCodeScan/Analyzers/Taint/CodeEvaluation/CSharpCodeEvaluation.cs [hunk 3]
```diff
@@ -778,6 +783,70 @@ private bool CheckPrecondition(IReadOnlyDictionary<int, object> condition, bool
...
+            for(var i = 0; i < ps.Length; i++)
+            {
+                var p = ps[i];
+                if(p.HasExplicitDefaultValue)
+                {
+                    vals[i] = p.ExplicitDefaultValue?.ToString();
```
Comment (JarLob) [inline SecurityCodeScan/Analyzers/Taint/CodeEvaluation/CSharpCodeEvaluation.cs]: Again for performance it is better to avoid string allocation. Could the `object` be used instead?

File: SecurityCodeScan/Analyzers/Taint/CodeEvaluation/CSharpCodeEvaluation.cs [hunk 4]
```diff
@@ -778,6 +783,70 @@ private bool CheckPrecondition(IReadOnlyDictionary<int, object> condition, bool
...
+            {
+                var destIx = methodSymbol?.FindArgumentIndex(lexicalIx, arg) ?? lexicalIx;
+
+                var val = state.AnalysisContext.SemanticModel.GetConstantValue(arg.Expression);
+                if (val.HasValue)
+                {
+                    if (destIx >= vals.Length)
+                        return false;
+
+                    vals[destIx] = val.Value?.ToString();
```
Comment (JarLob) [inline SecurityCodeScan/Analyzers/Taint/CodeEvaluation/CSharpCodeEvaluation.cs]: Again for performance it is better to avoid string allocation. Could the `object` be used instead?

File: SecurityCodeScan/Analyzers/Taint/CodeEvaluation/CSharpCodeEvaluation.cs [hunk 5]
```diff
@@ -778,6 +783,70 @@ private bool CheckPrecondition(IReadOnlyDictionary<int, object> condition, bool
...
+            var lexicalIx = 0;
+            foreach (var arg in args)
+            {
+                var destIx = methodSymbol?.FindArgumentIndex(lexicalIx, arg) ?? lexicalIx;
+
+                var val = state.AnalysisContext.SemanticModel.GetConstantValue(arg.Expression);
+                if (val.HasValue)
+                {
+                    if (destIx >= vals.Length)
+                        return false;
```
Comment (JarLob) [inline SecurityCodeScan/Analyzers/Taint/CodeEvaluation/CSharpCodeEvaluation.cs]: When does it happen? With `params`? Could you add a test? If this is impossible situation I would rather throw.

File: SecurityCodeScan/Analyzers/Taint/CodeEvaluation/CSharpCodeEvaluation.cs [hunk 6]
```diff
@@ -778,6 +783,70 @@ private bool CheckPrecondition(IReadOnlyDictionary<int, object> condition, bool
...
+
+                    vals[destIx] = val.Value?.ToString();
+                }
+
+                lexicalIx++;
+            }
+
+            foreach (var kv in condition)
+            {
+                var ix = int.Parse((string)kv.Key);
```
Comment (JarLob) [inline SecurityCodeScan/Analyzers/Taint/CodeEvaluation/CSharpCodeEvaluation.cs]: Again, remember about optimized internal configuration structure. This can be done only once during config loading.

File: SecurityCodeScan/Analyzers/Taint/CodeEvaluation/CSharpCodeEvaluation.cs [hunk 7]
```diff
@@ -778,6 +783,70 @@ private bool CheckPrecondition(IReadOnlyDictionary<int, object> condition, bool
...
+
+                if (ix >= vals.Length)
+                    return false;
```
Comment (JarLob) [inline SecurityCodeScan/Analyzers/Taint/CodeEvaluation/CSharpCodeEvaluation.cs]: Invalid configuration? I would throw and let the user know his config is wrong.

File: SecurityCodeScan/Analyzers/Taint/CodeEvaluation/CSharpCodeEvaluation.cs [hunk 8]
```diff
@@ -778,6 +783,70 @@ private bool CheckPrecondition(IReadOnlyDictionary<int, object> condition, bool
...
+
+                var expectedVal = ((IReadOnlyDictionary<object, object>)kv.Value)["Value"];
+                var codeVal = vals[ix];
+
+                if (codeVal == null)
+                    return false;
+
+                if (!codeVal.Equals(expectedVal.ToString()))
```
Comment (JarLob) [inline SecurityCodeScan/Analyzers/Taint/CodeEvaluation/CSharpCodeEvaluation.cs]: Compare objects instead?

**Comment by kevin-montrose** [comment]
Closing this in favor of #137, which is against master.
